python/example_code/cognito/cognito_idp_actions.py

```diff
@@ -0,0 +1,405 @@
+        that can be used to get AWS credentials from Amazon Cognito.
+
+        :param user_name: The name of the user who is signing in.
+        :param session: Session information returned from a previous call to initiate
+                        authentication.
+        :param mfa_code: A code generated by the associated MFA application.
+        :return: The result of the authentication. When successful, this contains an
+                 access token for the user.
+        """
+        try:
+            kwargs = {
+                'UserPoolId': self.user_pool_id,
+                'ClientId': self.client_id,
+                'ChallengeName': 'SOFTWARE_TOKEN_MFA', 'Session': session,
+                'ChallengeResponses': {
+                    'USERNAME': user_name, 'SOFTWARE_TOKEN_MFA_CODE': mfa_code}}
+            if self.client_secret is not None:
+                kwargs['ChallengeResponses']['SECRET_HASH'] = self._secret_hash(user_name)
+            response = self.cognito_idp_client.admin_respond_to_auth_challenge(**kwargs)
+            auth_result = response['AuthenticationResult']
+        except ClientError as err:
+            if err.response['Error']['Code'] == 'ExpiredCodeException':
+                logger.warning(
+                    "Your MFA code has expired or has been used already. You may have "
+                    "to wait a few seconds until your app shows you a new code.")
+            else:
+                logger.error(
+                    "Couldn't respond to mfa challenge for %s. Here's why: %s: %s", user_name,
+                    err.response['Error']['Code'], err.response['Error']['Message'])
+                raise
+        else:
+            return auth_result
+    # snippet-end:[python.example_code.cognito-idp.AdminRespondToAuthChallenge]
+
+    # snippet-start:[python.example_code.cognito-idp.ConfirmDevice]
+    def confirm_mfa_device(
+            self, user_name, device_key, device_group_key, device_password, access_token,
+            aws_srp):
+        """
+        Confirms an MFA device to be tracked by Amazon Cognito. When a device is
+        tracked, its key and password can be used to sign in without requiring a new
+        MFA code from the MFA application.
+
+        :param user_name: The user that is associated with the device.
+        :param device_key: The key of the device, returned by Amazon Cognito.
+        :param device_group_key: The group key of the device, returned by Amazon Cognito.
+        :param device_password: The password that is associated with the device.
+        :param access_token: The user's access token.
+        :param aws_srp: A class that helps with Secure Remote Password (SRP)
+                        calculations. The scenario associated with this example uses
+                        the warrant package.
+        :return: True when the user must confirm the device; otherwise, False. When
+                 False, the device is automatically confirmed and tracked.
```

Comment:
:return: True when the user must confirm the device. Otherwise, False. When
                 False, the device is automatically confirmed and tracked.

Reply:
fixed